metadata/src/main/java/org/apache/kafka/controller/FeatureControlManager.java

```diff
@@ -177,9 +178,19 @@ ControllerResult<Map<String, ApiError>> updateFeatures(
         TreeMap<String, ApiError> results = new TreeMap<>();
         List<ApiMessageAndVersion> records =
                 BoundedList.newArrayBacked(MAX_RECORDS_PER_USER_OP);
+
+        Map<String, Short> proposedUpdatedVersions = new HashMap<>();
+        finalizedVersions.forEach(proposedUpdatedVersions::put);
+        proposedUpdatedVersions.put(MetadataVersion.FEATURE_NAME, metadataVersion.get().featureLevel());
+        updates.forEach(proposedUpdatedVersions::put);
+
         for (Entry<String, Short> entry : updates.entrySet()) {
-            results.put(entry.getKey(), updateFeature(entry.getKey(), entry.getValue(),
-                upgradeTypes.getOrDefault(entry.getKey(), FeatureUpdate.UpgradeType.UPGRADE), records));
+            ApiError error = updateFeature(entry.getKey(), entry.getValue(),
+                upgradeTypes.getOrDefault(entry.getKey(), FeatureUpdate.UpgradeType.UPGRADE), records, proposedUpdatedVersions);
+            if (!error.error().equals(Errors.NONE)) {
+                return ControllerResult.of(Collections.emptyList(), Collections.singletonMap(entry.getKey(), error));
+            }
+            results.put(entry.getKey(), error);
```

Comment:
Could we just return `ControllerResult<ApiError>` for `updateFeatures`?

Reply:
I suppose we can revise this.

Reply:
Actually, need to take a closer look at how this is implemented.

Reply:
This worked out nicely 👍